### DISCUSSION TITLE
Report: Good Japanese ASR results with Fast Conformer

### DISCUSSION THREAD

**Discussion by fujimotos** [NVIDIA-NeMo/NeMo#8473] [Public Models]
Hi, we are Speech research team from Reazon Human Interaction Lab. [^1]

[^1]: A research division of Reazon Holdings, a Tokyo-based infotech company.

Recently, we experimented with [Fast Conformer](https://arxiv.org/abs/2305.05084) using Japanese datasets,
and could confirm that it delivers excellent performance. So we want to
share our findings.

We hope this post is interesting to NeMo team.

## Method

We trained a Fast Conformer model using [NeMo framework](https://github.com/nvidia/nemo) with the
following configuration:

 * Subword-based RNN-T model with 619M parameters.

 * Encoder uses Longformer-style attention with limited context size of `[128, 128]`.
   We enabled local attention from the beginning of training.

 * Decoder has a vocabulary space of 3000 tokens constructed by
   [SentencePiece](https://github.com/google/sentencepiece) unigram tokenizer.

We trained this model on our [ReazonSpeech v2.0 corpus](https://huggingface.co/datasets/reazon-research/reazonspeech), which provides
35,000 hours of diverse Japanese speech. [^2]

[^2]: For more details about this dataset, refer to [our paper](http://research.reazon.jp/_static/reazonspeech_nlp2023.pdf) last year.

## Result (Model)

We released our model on Hugging Face under Apache License 2.0.

https://huggingface.co/reazon-research/reazonspeech-nemo-v2

### 1) Inference Speed vs Accuracy 

This graph compares accuracy and inference speed of Japanese ASR models.

<img src="https://github.com/NVIDIA/NeMo/assets/8974561/56c9081d-7579-48ad-bc18-7691aac746eb" width="600">

* The x-axis represents the speed of inference, expressed as real-time
  factor.

* The y-axis represents the character error rate on JSUT-book dataset
  (1 hour of Japanese read speech).

As you can see, the Fast Conformer model archives better inference accuracy
than Whisper v1/2/3, while being as fast as Whisper Tiny.

### 2) Robustness

Here is another graph that compares Japanese ASR models on public corpus:

<img src="https://github.com/NVIDIA/NeMo/assets/8974561/b310b0b4-54ab-4614-bf96-7bbef58c9af2" width="650">

So, the Fast Conformer model shows robust performance across multiple datasets.

**Comment by titu1994** [comment]
This is excellent news ! Very glad Fast Conformer achieved strong results while being much more efficient than the largest Whisper models.

**Comment by galv** [comment]
What batch size are you using for RTF? Is the batch size 1?

For RNN-T greedy decoding inference, we have noticed that the majority of the time is spent in the greedy decoder. My PR https://github.com/NVIDIA/NeMo/pull/8191 which proposes a fix by reducing the time spent waiting on the CPU in each iteration will go in shortly. At batch size 16, for a 600 million parameter model, I see a 3.125x speed up. There is also complementary work from @artbataev in #8286 to reduce the number of iterations done in decoding, though we have yet to combine the two efforts into one implementation using both.

**Reply by fujimotos** [reply]
@galv We did a bit of research on decoding strategies implemented in NeMo,
and we found that the strategy choice has significant effects on inference results.

The following graph compares Greedy, ALSD, TSD, and mAES:

<img src='https://github.com/NVIDIA/NeMo/assets/8974561/7e050317-6886-48e1-88ea-9c66300f2ce9' width=600>

All the measurements were done with batch size 1. For reference, the dict
below shows the decoding parameter used in our testing:

```python
{'strategy': '<alsd, tsd, greedy or beam>',
 'beam': {'alsd_max_target_len': 1.0,
          'beam_size': 4,
          'preserve_alignments': False,
          'return_best_hypothesis': True,
          'score_norm': True,
          'search_type': 'default',
          'softmax_temperature': 1.0},
 'greedy': {'max_symbols': 10}}
```

Based on this result, we ended up using ALSD. We found it produced
the best inference quality.

> What batch size are you using for RTF? Is the batch size 1?

As mentioned above, the batch size was 1 (both for NeMo and Whisper).

> For RNN-T greedy decoding inference, we have noticed that the majority of the time is spent in the greedy decoder.

We could confirm this. Currently the majority of inference time is spent on
beam search, and it's CPU bound (especially for longform audio test sets).

**Comment by yuekaizhang** [comment]
> ### 1) Inference Speed vs Accuracy
> This graph compares accuracy and inference speed of Japanese ASR models.
> 
> <img alt="" width="600" src="https://private-user-images.githubusercontent.com/8974561/306653235-56c9081d-7579-48ad-bc18-7691aac746eb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NjY3MjAsIm5iZiI6MTcwODU2NjQyMCwicGF0aCI6Ii84OTc0NTYxLzMwNjY1MzIzNS01NmM5MDgxZC03NTc5LTQ4YWQtYmMxOC03NjkxYWFjNzQ2ZWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMDE0NzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTljZGE2OTRhMDZkZGI1MTQxMGE4NzU5NTBiYTg3NzY5NjcyZWU0ODU3YjMyYTJlYzY3ZDQ2MmQ4ODQ0YzlmMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.yqOn02_med7UcvZjsV6SBXWQqBFCNBHj8NXBLnsOX0E">
> * The x-axis represents the speed of inference, expressed as real-time
>   factor.
> * The y-axis represents the character error rate on JSUT-book dataset
>   (1 hour of Japanese read speech).
> 
> As you can see, the Fast Conformer model archives better inference accuracy than Whisper v1/2/3, while being as fast as Whisper Tiny.
> 
> ### 2) Robustness
> Here is another graph that compares Japanese ASR models on public corpus:
> 
> <img alt="" width="650" src="https://private-user-images.githubusercontent.com/8974561/306653109-b310b0b4-54ab-4614-bf96-7bbef58c9af2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NjY3MjAsIm5iZiI6MTcwODU2NjQyMCwicGF0aCI6Ii84OTc0NTYxLzMwNjY1MzEwOS1iMzEwYjBiNC01NGFiLTQ2MTQtYmY5Ni03YmJlZjU4YzlhZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMDE0NzAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2UxYzcwM2Q1NmRhNTA5MTVjMzg3NzU1NjFlNzcyYWFmYTkxNDQ3YjQ2NDViNWZmYzU1MWNjNGJjNzcyZmY3ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.PRqjGm7jKW_1fCVMvVlzGhZYdRbpAUWmEcAzchwORl0">

@fujimotos I was wondering whether the above results of Whisper be without any fine-tuning? Do we have WER experimental results for Whisper after fine-tuning on the ReazonSpeech v2.0 corpus?

**Reply by fujimotos** [reply]
@yuekaizhang These results are of official Whisper models (no finetuning).
We have some idea on finetuning Whisper, but not working on it yet.

**Comment by LostnD** [comment]
doesn't work I got this below error 

```
[NeMo I 2024-11-17 22:38:15 save_restore_connector:263] Model EncDecRNNTBPEModel was successfully restored from C:\Users\Andriken\.cache\huggingface\hub\models--reazon-research--reazonspeech-nemo-v2\snapshots\33693408be76b7cba9fd4a7546a0a8772430211b\reazonspeech-nemo-v2.nemo.
[NeMo I 2024-11-17 22:38:15 align:250] Flag use_local_attention is set to True => will try to use local attention for model if it allows it
Error executing job with overrides: ['pretrained_name=reazon-research/reazonspeech-nemo-v2', 'manifest_filepath=manifest_with_text.json', 'output_dir=data/nfa_output/', 'additional_segment_grouping_separator=|', 'ass_file_config.vertical_alignment=bottom', 'ass_file_config.text_already_spoken_rgb=[66,245,212]', 'ass_file_config.text_being_spoken_rgb=[242,222,44]', 'ass_file_config.text_not_yet_spoken_rgb=[223,242,239]']
Traceback (most recent call last):
  File "D:\cu\Nemo\NeMo\tools\nemo_forced_aligner\align.py", line 256, in main
    raise NotImplementedError(
NotImplementedError: Model is not an instance of NeMo EncDecCTCModel or ENCDecHybridRNNTCTCModel. Currently only instances of these models are supported
```

my command was this 

`(D:\cu\Nemo\nemoenv) PS D:\cu\Nemo> python NeMo/tools/nemo_forced_aligner/align.py pretrained_name="reazon-research/reazonspeech-nemo-v2" manifest_filepath=manifest_with_text.json output_dir=data/nfa_output/ additional_segment_grouping_separator="|" ass_file_config.vertical_alignment="bottom" ass_file_config.text_already_spoken_rgb=[66,245,212] ass_file_config.text_being_spoken_rgb=[242,222,44] ass_file_config.text_not_yet_spoken_rgb=[223,242,239]`

**Reply by bfreuden** [reply]
Did you manage to solve the issue?

**Comment by PhamDangNguyen** [comment]
I get an error liked "RuntimeError: Internal: src/trainer_interface.cc(582) [(static_cast<int>(required_chars_.size() + meta_pieces_.size())) <= (trainer_spec_.vocab_size())] Vocabulary size is smaller than required_chars. 3072 vs 4512. Increase vocab_size or decrease character_coverage with --character_coverage option." ?? How can i get the number of vocab is 3072 with my dataset? (my dataset is 5000 hours ReazonSpeech larg)

**Reply by fujimotos** [reply]
If you are trying to finetune an existing NeMo model, you may need to
override the vocaburary. Refer to the following tutorial for an example:

https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb#scrollTo=gSI6t9dgSOxj

The best option will be keeping the original vocaburary, though.
It makes the transfer learning more streight-forward.

**Reply by PhamDangNguyen** [reply]
thanks, I got it

**Comment by PhamDangNguyen** [comment]
I try training model fast-conformer larg 120M from scratch but it's not converge, this is my config:
# It contains the default values for training a Fast Conformer-CTC ASR model, large size (~120M) with CTC loss and sub-word encoding.
name: "FastConformer-CTC-BPE"

model:
  sample_rate: 16000
  log_prediction: true # enables logging sample predictions in the output during training
  ctc_reduction: 'mean_volume'
  skip_nan_grad: false

  train_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/train_remove_func_18.json
    sample_rate: ${model.sample_rate}
    batch_size: 40 # you may increase batch_size if your memory allows
    shuffle: true
    num_workers: 32
    pin_memory: true
    max_duration: 21 # it is set for LibriSpeech, you may need to update it for your dataset
    min_duration: 0.3
    # tarred datasets
    is_tarred: false
    tarred_audio_filepaths: null
    shuffle_n: 2048
    # bucketing params
    bucketing_strategy: "fully_randomized"
    bucketing_batch_size: null

  validation_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/test_remove_func.json
    sample_rate: ${model.sample_rate}
    batch_size: 16 # you may increase batch_size if your memory allows
    shuffle: false
    use_start_end_token: false
    num_workers: 16
    pin_memory: true

  test_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/test_remove_func.json
    sample_rate: ${model.sample_rate}
    batch_size: 16 # you may increase batch_size if your memory allows
    shuffle: false
    use_start_end_token: false
    num_workers: 16
    pin_memory: true

  tokenizer:
    dir: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/dict_N/tokenizer_spe_bpe_v3072  # path to directory which contains either tokenizer.model (bpe) or vocab.txt (wpe)
    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: 16000
    normalize: "per_feature"
    window_size: 0.025
    window_stride: 0.01
    window: "hann"
    features: 80
    n_fft: 512
    log: true
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0
    pad_value: 0.0

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2 # set to zero to disable it
    # you may use lower time_masks for smaller models to have a faster convergence
    # time_masks: 10 # set to zero to disable it
    time_masks: 0 
    freq_width: 27
    time_width: 0.05

  encoder:
    _target_: nemo.collections.asr.modules.ConformerEncoder
    feat_in: ${model.preprocessor.features}
    feat_out: -1 # you may set it if you need different output size other than the default d_model
    n_layers: 19
    d_model: 512

    # Sub-sampling params
    subsampling: dw_striding # vggnet, striding, stacking or stacking_norm, dw_striding
    subsampling_factor: 8 # must be power of 2 for striding and vggnet
    subsampling_conv_channels: 256 # -1 sets it to d_model
    causal_downsampling: false

    # Feed forward module's params
    ff_expansion_factor: 4

    # Multi-headed Attention Module's params
    self_attention_model: rel_pos # rel_pos or abs_pos
    n_heads: 8 # may need to be lower for smaller d_models
    # [left, right] specifies the number of steps to be seen from left and right of each step in self-attention
    att_context_size: [-1, -1] # -1 means unlimited context
    att_context_style: regular # regular or chunked_limited
    xscaling: true # scales up the input embeddings by sqrt(d_model)
    untie_biases: true # unties the biases of the TransformerXL layers
    pos_emb_max_len: 5000

    # Convolution module's params
    conv_kernel_size: 9
    conv_norm_type: 'batch_norm' # batch_norm or layer_norm or groupnormN (N specifies the number of groups)
    # conv_context_size can be"causal" or a list of two integers while conv_context_size[0]+conv_context_size[1]+1==conv_kernel_size
    # null means [(kernel_size-1)//2, (kernel_size-1)//2], and 'causal' means [(kernel_size-1), 0]
    conv_context_size: null

    ### regularization
    dropout: 0.1 # The dropout used in most of the Conformer Modules
    dropout_pre_encoder: 0.1 # The dropout used before the encoder
    dropout_emb: 0.0 # The dropout used for embeddings
    dropout_att: 0.1 # The dropout for multi-headed attention modules

    # set to non-zero to enable stochastic depth
    stochastic_depth_drop_prob: 0.0
    stochastic_depth_mode: linear  # linear or uniform
    stochastic_depth_start_layer: 1

  decoder:
    _target_: nemo.collections.asr.modules.ConvASRDecoder
    feat_in: null
    num_classes: -1
    vocabulary: []


  interctc:
    loss_weights: []
    apply_at_layers: []

  optim:
    name: adamw
    lr: 1e-3
    betas: [0.9, 0.98]
    weight_decay: 1e-3

    # scheduler setup
    sched:
      name: CosineAnnealing
      warmup_ratio: 0.1
      min_lr: 5e-4

trainer:
  devices: 3 # number of GPUs, -1 would use all available GPUs
  num_nodes: 1
  max_epochs: 1000
  max_steps: -1 # computed at runtime if not set
  val_check_interval: 1.0 # Set to 0.25 to check 4 times per epoch, or an int for number of iterations
  accelerator: auto
  strategy: ddp
  accumulate_grad_batches: 1
  gradient_clip_val: 0.0
  precision: 32 # 16, 32, or bf16
  log_every_n_steps: 10  # Interval of logging.
  enable_progress_bar: True
  num_sanity_val_steps: 0 # number of steps to perform validation steps for sanity check the validation process before starting the training, setting to 0 disables it
  check_val_every_n_epoch: 1 # number of evaluations on validation every n epochs
  sync_batchnorm: true
  enable_checkpointing: False  # Provided by exp_manager
  logger: false  # Provided by exp_manager
  benchmark: false # needs to be false for models with variable-length speech input as it slows down training

exp_manager:
  exp_dir: null
  name: ${name}
  create_tensorboard_logger: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    monitor: "val_wer"
    mode: "min"
    save_top_k: 40
    always_save_nemo: True # saves the checkpoints as nemo files instead of PTL checkpoints

  resume_from_checkpoint: /home/team_voice/STT_pdnguyen/Backup_version_train_L40S/fast_conformer_ja_larg/25_12_2/nemo_experiments/FastConformer-CTC-BPE/2025-02-11_08-17-59/checkpoints/FastConformer-CTC-BPE--val_wer=1.0000-epoch=3.ckpt
  resume_if_exists: true
  resume_ignore_no_checkpoint: true

  create_wandb_logger: false
  wandb_logger_kwargs:
    name: null
    project: null
Can you give me some advice? Also, if you have a pre-trained JA model with 120M parameters, please release it for me. Thanks

**Reply by Houss3m** [reply]
was it predicting blank (nothing) during training for you?

**Comment by vanIvan** [comment]
@fujimotos thank you for opensourcing such a good model! 

Could you please share training config - recipe for your model? I was not able to find anywhere (reazonspeech webpage / hugginface artefacts / here)

And may be you have results for training models on smaller sets of reazonspeech dataset like M or L subsets, that you can share?

**Reply by fujimotos** [reply]
@vanIvan Actually I have left the lab a few months ago (got to work on my own stuff).

So you should contact the current lab members. I think you can reach them through
GitHub issue on ReazonSpeech project.
